flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/OneSlotPerExecutionSlotAllocator.java

```diff
@@ -0,0 +1,260 @@
+				slotProfileFuture.thenApply(
+					slotProfile -> createPhysicalSlotRequest(slotRequestId, slotProfile));
+			physicalSlotRequestFutures.add(physicalSlotRequestFuture);
+		}
+
+		return physicalSlotRequestFutures;
+	}
+
+	private PhysicalSlotRequest createPhysicalSlotRequest(
+			final SlotRequestId slotRequestId,
+			final SlotProfile slotProfile) {
+		return new PhysicalSlotRequest(slotRequestId, slotProfile, slotProviderStrategy.willSlotBeOccupiedIndefinitely());
+	}
+
+	private void allocateSlotsForAssignments(
+			final List<CompletableFuture<PhysicalSlotRequest>> physicalSlotRequestFutures,
+			final List<SlotExecutionVertexAssignment> slotExecutionVertexAssignments,
+			final Map<ExecutionVertexID, SlotRequestId> executionVertexSlotRequestIds) {
+
+		final Map<ExecutionVertexID, SlotExecutionVertexAssignment> vertexToAssignmentMapping =
+			slotExecutionVertexAssignments.stream()
+				.collect(Collectors.toMap(SlotExecutionVertexAssignment::getExecutionVertexId, Function.identity()));
+
+		final Map<SlotRequestId, ExecutionVertexID> slotToVertexMapping = new HashMap<>();
+		for (Map.Entry<ExecutionVertexID, SlotRequestId> entry : executionVertexSlotRequestIds.entrySet()) {
+			slotToVertexMapping.put(entry.getValue(), entry.getKey());
+		}
+
+		FutureUtils.combineAll(physicalSlotRequestFutures)
+			.thenCompose(physicalSlotRequests -> slotProviderStrategy.allocatePhysicalSlots(physicalSlotRequests))
+			.thenAccept(physicalSlotRequestResults -> {
+				for (PhysicalSlotRequest.Result result : physicalSlotRequestResults) {
+					final SlotRequestId slotRequestId = result.getSlotRequestId();
+					final ExecutionVertexID executionVertexId = slotToVertexMapping.get(slotRequestId);
+					try {
+						final LogicalSlot logicalSlot = allocateSingleLogicalSlotFromPhysicalSlot(
+							slotRequestId,
+							result.getPhysicalSlot());
+						vertexToAssignmentMapping.get(executionVertexId).getLogicalSlotFuture().complete(logicalSlot);
+					} catch (Exception ex) {
+						throw new CompletionException(ex);
+					}
+				}
+			})
+			.exceptionally(ex -> {
+				executionVertexSlotRequestIds.values().stream()
+					.forEach(slotRequestId -> {
+						final ExecutionVertexID executionVertexId = slotToVertexMapping.get(slotRequestId);
+						vertexToAssignmentMapping.get(executionVertexId).getLogicalSlotFuture().completeExceptionally(ex);
+					});
+				return null;
+			});
+	}
+
+	private LogicalSlot allocateSingleLogicalSlotFromPhysicalSlot(
+			final SlotRequestId slotRequestId,
+			final PhysicalSlot physicalSlot) throws FlinkException {
+
+		final SingleLogicalSlot singleTaskSlot = new SingleLogicalSlot(
+			slotRequestId,
+			physicalSlot,
+			null,
+			Locality.UNKNOWN,
+			slotOwner);
+
+		if (physicalSlot.tryAssignPayload(singleTaskSlot)) {
+			return singleTaskSlot;
+		} else {
+			throw new FlinkException("Could not assign payload to allocated slot " + physicalSlot.getAllocationId() + '.');
```

Comment:
All pending requested and allocated slots will be released on this exception.
This is because `slotProviderStrategy.cancelSlotRequest()` will be invoked.